### PR TITLE
Set NotNull constraints for member references in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to https://github.com/SonarSource/sonar-dotnet/issues/5860

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 1]
```diff
@@ -567,5 +567,133 @@ public void Main<T>() where T : new()
...
+Tag(""Before"", array);
+array[42] = 42;
+Tag(""After"", array);
+
+int[] UntrackedSymbol() => new[] { 42 };";
+            var validator = SETestContext.CreateCS(code, ", int[] array").Validator;
+            validator.ValidateContainsOperation(OperationKind.ArrayElementReference);
+            validator.ValidateTag("Before", x => x.Should().BeNull());
+            validator.ValidateTag("After", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:697]: We have get and set operations. Would be nice to have a data test for all the other operators
* Invocation
* Event subscription
* Post and Pre operators (++)
* Casts

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:697]: A test for dynmic binding is missing
```cs
, dynamic arg
arg.BlaBla();
arg[0] = 0;
arg.Prop = 42;
// ...
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some test are missing IMHO.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 2]
```diff
@@ -567,5 +567,133 @@ public void Main<T>() where T : new()
...
+UntrackedSymbol().Property = 42;
+Tag(""BeforeProperty"", arg);
+Tag(""BeforeDictionary"", dictionary);
+Tag(""BeforeIndexer"", indexer);
+arg.Property = 42;
+dictionary[42] = 42;
+indexer[42] = 42;
+Tag(""AfterProperty"", arg);
+Tag(""AfterDictionary"", dictionary);
+Tag(""AfterIndexer"", indexer);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:654]: I don't get this. The functionality of how symbols are tracked is a responsibility of another UT that we don't need to re-do here.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:697]: We are not field-sensitive, we have no symbolic value for `arg.Joe` so we can't build constraints for that (and anything inside that).

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:697]: Invocation is another PR.
I'll check how event subscription can be done.
I don't see a reason to build constraints on `++`, operators does not guarantee `NotNull`
Cast doesn't guarantee `NotNull` either.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:697]: I'll do the dynamic stuff later, if I have some time. That one is not that much valuable.

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:839]: > We are not field-sensitive, we have no symbolic value for arg.Joe so we can't build constraints for that (and anything inside that).

That is unfortunate and very limiting. Are there plans to overcome this?

**Review by martin-strecker-sonarsource — APPROVED** [review]
Can we overcome the nesting limitation?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:839]: It's an idea for `the next big thing`, nothing is specified.

**Pull request merged**